File: core/grammar.py
import os
import time
from openai import OpenAI
from tqdm import tqdm

from .text_normalize import normalize_text, fix_punct_local, needs_qwen
from .cache import CorrectionCache
from .config import PROFILES

_API_KEY = os.environ.get("OMLX_API_KEY", "pippopippo")
_client = OpenAI(base_url="http://127.0.0.1:8000/v1", api_key=_API_KEY)

_SYS = (
    "Sei un correttore ortografico automatico per sottotitoli video italiani. "
    "INPUT: una frase breve, possibilmente con errori di battitura, "
    "accordo grammaticale o trascrizione automatica. "
    "OUTPUT: restituisci SOLO la frase corretta, senza spiegazioni, senza virgolette, "
    "senza prefissi come Correzione:, senza aggiungere frasi nuove, "
    "senza completare pensieri lasciati volutamente incompleti. "
    "NON aggiungere parole a meno che non siano strettamente necessarie per la grammatica. "
    "NON togliere parole. NON cambiare il significato. NON punteggiare alla fine. "
    "Se la frase è già corretta, restituiscila identica."
)

_cache = CorrectionCache()


def _validate(output: str, original: str) -> str:
    stripped = output.strip().strip('"').strip("'").strip()
    words_out = stripped.split()
    words_in = original.strip().split()
    if len(words_out) > len(words_in) * 2:
        return ""
    if any(c in stripped for c in ["\n", "→", "Correzione", "Nota", "**"]):
        return stripped.split("\n")[0]
    if ":" in stripped and stripped.count(" ") < 6:
        return stripped.split(":")[-1].strip()
    return stripped


def _correct_text(text: str, model: str) -> str:
    if not text.strip():
        return text
    try:
        r = _client.chat.completions.create(
            model=model,
            messages=[{"role": "system", "content": _SYS}, {"role": "user", "content": text}],
            temperature=0.0,
            max_tokens=len(text) * 3 + 30,
        )
        raw = r.choices[0].message.content.strip()
        out = _validate(raw, text)
        return out if out else text
    except Exception:
        return text


def _correct_with_cache(text: str, model: str) -> str:
    return _cache.get_or_correct(text, lambda t: _correct_text(t, model))


def _correct_batch(segments: list[dict], model: str) -> list[dict]:
    to_correct = [s for s in segments if needs_qwen(s["text"])]
    cache_hits = 0

    for seg in to_correct:
        cached = _cache.get(seg["text"])
        if cached is not None:
            seg["text"] = cached
            cache_hits += 1

    to_query = [s for s in to_correct if s["text"] == s.get("_original", s["text"])]

    for seg in to_query:
        corrected = _correct_text(seg["text"], model)
        if corrected:
            _cache.set(seg["text"], corrected)
            seg["text"] = corrected

    return segments


def correct_segments(
    segments: list[dict],
    model: str,
    profile: dict | None = None,
    on_corrected=None,
) -> list[dict]:
    if profile is None:
        profile = PROFILES.get("conversational")

    corrected = []
    batch_size = 5

    if model:
        for i in range(0, len(segments), batch_size):
            batch = segments[i:i + batch_size]
            batch = _correct_batch(batch, model)
            corrected.extend(batch)
    else:
        corrected = [dict(s) for s in segments]

    for seg in corrected:
        seg["text"] = normalize_text(seg["text"])

    context: dict = {"prev_gap": 0.0}
    for i, seg in enumerate(corrected):
        next_seg = corrected[i + 1] if i + 1 < len(corrected) else None
        next_text = next_seg["text"] if next_seg else ""
        gap = (next_seg["start"] - seg["end"]) if next_seg else 2.0

        seg["text"] = fix_punct_local(seg["text"], next_text=next_text, gap_sec=gap, profile=profile, context=context)

        if seg["text"] and seg["text"][0].islower():
            seg["text"] = seg["text"][0].upper() + seg["text"][1:]

        context["prev_gap"] = gap

        if on_corrected:
            on_corrected({"start": seg["start"], "end": seg["end"], "text": seg["text"]})

    return corrected

File: src-tauri/src/services/grammar.rs
use videoforge_core::CorrectionCache;

static SYSTEM_PROMPT: &str = "Sei un correttore ortografico e di punteggiatura per sottotitoli video italiani.

Regole:
- Correggi errori di battitura, ortografia e accordo grammaticale.
- Aggiungi la punteggiatura mancante (punti, virgole, punti interrogativi).
- Se la frase termina, metti un punto. Se è una domanda, metti il punto interrogativo.
- Capitalizza la prima lettera della frase.
- Non spezzare le parole. Non dividere frasi. Non togliere pezzi di frase.
- Non parafrasare e non cambiare lo stile o il registro originale.
- Se la frase è già corretta e ben punteggiata, restituiscila identica.
- Restituisci SOLO il testo corretto, senza prefissi, spiegazioni o virgolette.";

pub struct GrammarService {
    base_url: String,
    api_key: String,
    cache: CorrectionCache,
}

impl GrammarService {
    pub fn new(base_url: &str, api_key: &str) -> Self {
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            api_key: if api_key.is_empty() {
                crate::services::env::EnvLoader::load()
                    .get("API_KEY")
                    .cloned()
                    .unwrap_or_default()
            } else {
                api_key.to_string()
            },
            cache: CorrectionCache::new(None),
        }
    }

    pub async fn list_models(&self) -> Vec<String> {
        let url = format!("{}/v1/models", self.base_url);
        let client = reqwest::Client::new();
        let resp = client
            .get(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .send()
            .await;

        let json: serde_json::Value = match resp {
            Ok(r) => match r.json().await {
                Ok(v) => v,
                Err(_) => return vec![],
            },
            Err(_) => return vec![],
        };

        json["data"]
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter_map(|m| m["id"].as_str().map(String::from))
                    .collect()
            })
            .unwrap_or_default()
    }

    pub async fn correct_segments(
        &self,
        segments: &[videoforge_core::Segment],
        model: &str,
    ) -> Vec<videoforge_core::Segment> {
        self.correct_batch(segments, model).await
    }

    async fn correct_batch(&self, segments: &[videoforge_core::Segment], model: &str) -> Vec<videoforge_core::Segment> {
        let mut result = segments.to_vec();

        for i in 0..result.len() {
            let text = result[i].text.clone();

            if let Some(cached) = self.cache.get(&text) {
                result[i].text = cached;
                continue;
            }

            let prev = if i > 0 { &result[i - 1].text } else { "" };
            let next = result.get(i + 1).map(|s| s.text.as_str()).unwrap_or("");

            if let Some(corrected) = self.correct_text_with_context(&text, prev, next, model).await {
                let validated = validate_output(&corrected, &text);
                if !validated.is_empty() {
                    self.cache.set(&text, &validated);
                    result[i].text = validated;
                }
            }
        }

        result
    }

    async fn correct_text_with_context(&self, text: &str, prev: &str, next: &str, model: &str) -> Option<String> {
        let url = format!("{}/v1/chat/completions", self.base_url);
        let client = reqwest::Client::new();

        let mut messages = vec![
            serde_json::json!({"role": "system", "content": SYSTEM_PROMPT}),
        ];

        if !prev.is_empty() {
            messages.push(serde_json::json!({"role": "user", "content": format!("FRASE PRECEDENTE: {}", prev)}));
            messages.push(serde_json::json!({"role": "assistant", "content": "OK."}));
        }

        messages.push(serde_json::json!({"role": "user", "content": format!("CORREGGI: {}", text)}));

        if !next.is_empty() {
            messages.push(serde_json::json!({"role": "user", "content": format!("(dopo continua con: {})", next)}));
        }

        let body = serde_json::json!({
            "model": model,
            "messages": messages,
            "temperature": 0.0,
            "max_tokens": text.len() * 3 + 50,
        });

        let resp = client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .ok()?;

        let json: serde_json::Value = resp.json().await.ok()?;
        let content = json["choices"][0]["message"]["content"].as_str()?;
        Some(content.trim().to_string())
    }
}

fn validate_output(output: &str, original: &str) -> String {
    let stripped = output.trim()
        .trim_matches(|c: char| c == '"' || c == '\'')
        .to_string();

    let words_out = stripped.split_whitespace().count();
    let words_in = original.split_whitespace().count();

    if words_out < words_in / 3 || words_out > words_in * 2 {
        return original.to_string();
    }

    if stripped.contains('\n') || stripped.contains("Correzione") || stripped.contains("Nota") || stripped.contains("**") {
        return original.to_string();
    }

    stripped
}
